feat: Adds the ability have internal endpoints listen on ksql.internal.listener

AlanConfluent opened the pull request:
### Description 
Allows for internal endpoints to listen on a separate address set by `ksql.internal.listener`.  This allows for separate treatment and better setup in an IaaS environment, opening up the possibility of port filtering.

It also is a precondition to treating intra-node traffic and external traffic differently, such as for implementing separate intra-node authentication and only authenticating users at the edges.  This will allow for better security (different roles for system and user requests) and performance improvements for pull queries, for example, since user auth can be checked just once.

If `ksql.internal.listener` is not set, it falls back to the existing behavior and has internal endpoints listen on `listeners`.

### Testing done 
mvn package.  Also tested with a cluster with authentication and verified `/heartbeat` and `/lag` works, along with pull queries and `list queries`.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

purplefox commented:
Hey @AlanConfluent I wonder if you could explain a bit more about the reasoning behind this. I think I haven't quite understood how this differs from ksql.advertised.listener, which aiui, is a mechanism designed for internal calls.

AlanConfluent commented:
> Hey @AlanConfluent I wonder if you could explain a bit more about the reasoning behind this? I think I haven't quite understood how this differs from ksql.advertised.listener, which aiui, is a mechanism designed for internal calls. Any chance you could elaborate a bit?

The idea is that we want to be able to separate internal calls from external calls.  Today, you can set `listeners` to some value and bind all endpoints to the same https://host:port (often something like https://0.0.0.0:9098), so the first purpose of this is to separate internal from external.

The second is to separate the binding vs broadcasting of the internal call address. `ksql.advertised.listener` allows you to broadcast a https://host:port to contact a node for internal calls (e.g. https://machine1234.internal.network.provider:9099). It's is very similar to `ksql.internal.listener` and for certain network setups, maybe they would be the same, but  `ksql.internal.listener` allows you to bind instead https://0.0.0.0:9099, for example, and be reachable on multiple interfaces/addresses.  Hopefully separating these two concepts allows for more flexibility.

In a cloud environment, you'd probably set all thread of these configs, with something similar to the examples I gave above.

purplefox commented:
> > Hey @AlanConfluent I wonder if you could explain a bit more about the reasoning behind this? I think I haven't quite understood how this differs from ksql.advertised.listener, which aiui, is a mechanism designed for internal calls. Any chance you could elaborate a bit?
> 
> The idea is that we want to be able to separate internal calls from external calls. Today, you can set `listeners` to some value and bind all endpoints to the same https://host:port (often something like https://0.0.0.0:9098), so the first purpose of this is to separate internal from external.
> 
> The second is to separate the binding vs broadcasting of the internal call address. `ksql.advertised.listener` allows you to broadcast a https://host:port to contact a node for internal calls (e.g. https://machine1234.internal.network.provider:9099). It's is very similar to `ksql.internal.listener` and for certain network setups, maybe they would be the same, but `ksql.internal.listener` allows you to bind instead https://0.0.0.0:9099, for example, and be reachable on multiple interfaces/addresses. Hopefully separating these two concepts allows for more flexibility.
> 

Maybe I'm just being slow, but I'm still not getting it :(

If you want to bind to both https://0.0.0.0:9098 and https://0.0.0.0:9099 then you can set up two listeners, and if you want the https://0.0.0.0:9099 listener to be used for internal calls you could set advertised listener to https://machine1234.internal.network.provider:9099 and the other nodes will end up connecting to the second listener. I'm sure I'm missing something here, but what does internal.listener buy you here?

AlanConfluent commented:
> If you want to bind to both https://0.0.0.0:9098 and https://0.0.0.0:9099 then you can set up two listeners, and if you want the https://0.0.0.0:9099 listener to be used for internal calls you could set advertised listener to https://machine1234.internal.network.provider:9099 and the other nodes will end up connecting to the second listener. I'm sure I'm missing something here, but what does internal.listener buy you here?

The above example works, and this is pretty much the situation today.  The issue is that if you want to set up your network to only allow external calls to come into the cluster, and block out internal call attempts, a common way of doing that is to do it by port.  You might disallow port 9099 for incoming requests, except within the private network containing your cluster.

In the above example, this type of port filtering isn't possible.  At least to the networking layer, you cannot tell the difference between public api calls and attempts at internal api calls.  You could make a proxy and configure it to whitelist various paths, but that's a more complex solution.

AlanConfluent commented on ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/Server.java:
```diff
@@ -65,26 +66,32 @@
     this.maxPushQueryCount = config.getInt(KsqlRestConfig.MAX_PUSH_QUERIES);
   }
 
+  // CHECKSTYLE_RULES.OFF: CyclomaticComplexity|NPathComplexity
```
Split up the method and removed the disabling.

purplefox commented:
> > If you want to bind to both https://0.0.0.0:9098 and https://0.0.0.0:9099 then you can set up two listeners, and if you want the https://0.0.0.0:9099 listener to be used for internal calls you could set advertised listener to https://machine1234.internal.network.provider:9099 and the other nodes will end up connecting to the second listener. I'm sure I'm missing something here, but what does internal.listener buy you here?
> 
> The above example works, and this is pretty much the situation today. The issue is that if you want to set up your network to only allow external calls to come into the cluster, and block out internal call attempts, a common way of doing that is to do it by port. You might disallow port 9099 for incoming requests, except within the private network containing your cluster.
> 
> In the above example, this type of port filtering isn't possible. At least to the networking layer, you cannot tell the difference between public api calls and attempts at internal api calls. You could make a proxy and configure it to whitelist various paths, but that's a more complex solution.

I think we should hop on a zoom, as I'm still not getting this :(

purplefox commented:
Just one observation for now- I think the motivation behind this is you want to drop auth on certain endpoints when accessed from the same internal network?
This seems risky from a security point of view - it will rely on the admin carefully choosing network interfaces for the internal listeners that aren't accessible to the outside world, or rely on network config to block access to specific interfaces from the outside world. A simple mistake in configuration could leave our servers wide open.
I'm guessing the reason you don't want to do auth for internal calls is for performance reasons (auth is slow and is currently done on each request) ? 
AIUI, currently internal calls are authed with the credentials of the public user who made the original public call. Instead, if executed internal calls with the credentials of the server itself (i.e. setup a new user "ksqldb-server", then we wouldn't need to drop auth for the internal endpoints, and we could cache authorisation for the internal traffic so there wouldn't be the auth hit each time. We could also cache the internal HTTP connection (have a shared ksqlDB client for the server internal traffic) so we wouldn't have connection setup time (network round trip, 2 x network round trip for TLS) each time. Just some thoughts.

AlanConfluent commented:
@purplefox Dropping auth for internal calls is partially for performance reasons.  The other reasons are that many internal calls are not done on behalf of the user or a given request since they are just system threads running in the background (e.g. heartbeating, lag).  We could still do auth on some of the endpoints, and we could use the config `KsqkRestConfig.AUTHENTICATION_SKIP_PATHS_CONFIG` to skip system endpoints or add them in `KsqlAuthorizationProviderHandler.PATHS_WITHOUT_AUTHORIZATION`.  Then the lack of auth is more self contained and new endpoints will automatically get auth by default.  Seems like a decent compromise.

Also, when `ksql.internal.listener` isn't set, `ServerVerticle` installs both `PortedEndpoints.setupEndpoints` and `PortedEndpoints.setupEndpointsInternal`.  Now, "/query" is installed in both and works fine.  I thought this might be an error, but it doesn't appear to be.  Should I ensure it's done just once?

@big-andy-coates Added functional tests and the additional internal endpoints for /query (pull queries) and /ksql (list queries, which uses the cluster discovery mechanism).  I'll do documentation tomorrow and update the thread when that's done.

AlanConfluent commented:
Please take another look @big-andy-coates @purplefox .

I believe @purplefox we've chatted about the motivation a bit more.

JimGalasyn commented on docs/operate-and-deploy/installation/server-config/config-reference.md:
```diff
@@ -415,6 +415,30 @@ The corresponding environment variable in the
+### ksql.advertised.listener
+
+The `ksql.advertised.listener` setting controls the address advertised within
+{{ site.kstreams }}, and shared with other ksqlDB nodes that make up the cluster.
```
```suggestion
{{ site.kstreams }} and is shared with other ksqlDB nodes that make up the cluster.
```

JimGalasyn commented on docs/operate-and-deploy/installation/server-config/config-reference.md:
```diff
@@ -415,6 +415,30 @@ The corresponding environment variable in the
 
+### ksql.advertised.listener
+
+The `ksql.advertised.listener` setting controls the address advertised within
```
```suggestion
The `ksql.advertised.listener` setting controls the address that is advertised within
```

JimGalasyn commented on docs/operate-and-deploy/installation/server-config/config-reference.md:
```diff
@@ -415,6 +415,30 @@ The corresponding environment variable in the
+This address is used to initiate intra-cluster requests such as forwarded pull queries,
+heartbeating, and lag reporting.
+
+If not set, the system will use the first of `listeners` to advertise and if this
```
```suggestion
If not set, the system will use the first host in `listeners` to advertise, and if this
```

JimGalasyn commented on docs/operate-and-deploy/installation/server-config/config-reference.md:
```diff
@@ -415,6 +415,30 @@ The corresponding environment variable in the
+heartbeating, and lag reporting.
+
+If not set, the system will use the first of `listeners` to advertise and if this
+is set to a wildcard address, then the system hostname is used.
```
```suggestion
is set to a wildcard address, the system hostname is used.
```

JimGalasyn commented on docs/operate-and-deploy/installation/server-config/config-reference.md:
```diff
@@ -415,6 +415,30 @@ The corresponding environment variable in the
+### ksql.internal.listener
+
+The `ksql.internal.listener ` setting controls the address bound for use by internal,
+intra-cluster endpoints.  These include forwarded pull queries, heartbeating, and lag reporting.
```
```suggestion
intra-cluster endpoints, which include forwarded pull queries, heartbeating, and lag reporting.
```

JimGalasyn commented on docs/operate-and-deploy/installation/server-config/config-reference.md:
```diff
@@ -415,6 +415,30 @@ The corresponding environment variable in the
+The `ksql.internal.listener ` setting controls the address bound for use by internal,
+intra-cluster endpoints.  These include forwarded pull queries, heartbeating, and lag reporting.
+
+If not set, the system will use `listeners` to expose internal endpoints.
```
```suggestion
If not set, the system will use `listeners` to expose internal endpoints.
```
```suggestion
If not set, the system uses `listeners` to expose internal endpoints.
```

JimGalasyn commented on docs/operate-and-deploy/installation/server-config/config-reference.md:
```diff
@@ -415,6 +415,30 @@ The corresponding environment variable in the
+
+If not set, the system will use `listeners` to expose internal endpoints.
+
+This setting is most often useful in a IaaS environment to separate external-facing
```
```suggestion
This setting is most often useful in an IaaS environment to separate external-facing
```

purplefox commented:
Hey @AlanConfluent what was the feedback from security?

big-andy-coates commented on ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/Server.java:
```diff
@@ -95,8 +102,13 @@ public synchronized void start() {
+        Optional.ofNullable(!internalListenUri.isPresent() ? internalEndpoints : null);
 
-    final int instances = config.getInt(KsqlRestConfig.VERTICLE_INSTANCES);
+    final int instances = 2; //config.getInt(KsqlRestConfig.VERTICLE_INSTANCES);
```
I'm not vertx, so can't really comment. @purplefox may have a view.

big-andy-coates commented:
Hey @AlanConfluent 

Sorry, I was getting confused as I'd forgotten `ksql.advertised.listener` was actually only used internally, as it's used to internal route pull query requests to the right node.

So we've currently got:
 * `listeners`: one or more endpoints to listen on.
 * `ksql.advertised.listener`: the listener for the node shared with other nodes. Currently defaults to `listeners`[0]`.

And you're proposing adding
 * `internal.listener`: a single endpoint to listen on for internal traffic. If not set, the internal calls are handled by `listeners` endpoints.

So now `ksql.advertised.listener`, if set, is an alias for `internal.listener`, if set.

This makes sense to me now.

Thanks for the test updates - however, I don't think I can see a test that ensures users can use the internal interfaces for non-internal calls.  This would be good to add.

The big issues I see are:

1. documentation: we have docs somewhere that explain how to setup the cluster, both with and without auth.   These will need updating.

2. auth testing: you're saying the current code will use the same auth method for internal and normal listeners.  That's ok for now, but I don't see a test that ensures auth is on internal listener, when configured. We'll need that.  I think this can be done in `AuthorizationFunctionalTest`.

Going forward, we'll likely need a better story around auth on the internal listeners.  This may be better done as a KLIP initially.  Likely, users won't want the same auth mech on the internal as the external.  

I think the current state is:

* Authentication:  currently ksql supports SSL, BASIC and/or a plugable mech.  Some paths can be configured to skip authentication.
* Authorisation: ksql doesn't currently internally do authorisation, and uses its own credentials when talking to external services.  Though in the future we want ksql to be able to impersonate users when talking to external services.

With this change the above will also apply to the internal interface too. However, I'm not sure how useful that is!  This shared auth mech would mean anyone that can call the public interface can also call the internal one. A more common pattern would be to allow the auth mech to be different on the two.

It would be good, (in a follow up), to split the configs to support different authentication for internal calls.  If the internal is not set it should default to the public auth.  A common setup may be to use ssl mutual auth for communication between nodes.

As @purplefox mentioned, we should definitely look to maintain a persistent connection between nodes, rather than creating them on the fly.  This will save a lot of overhead.

AlanConfluent commented:
> And you're proposing adding
> 
> * `internal.listener`: a single endpoint to listen on for internal traffic. If not set, the internal calls are handled by `listeners` endpoints.

Correct.

> So now `ksql.advertised.listener`, if set, is an alias for `internal.listener`, if set. If not set, `ksql.advertised.listeners` will default to `internal.listener`, if set, or `listeners`[0] otherwise.

I think technically, `ksql.advertised.listeners` will still default to `listeners`[0].  Maybe it should do as you're saying and fall back to `internal.listener` before that, but I wanted to keep it easy to reason about.

> 1. documentation: we have docs somewhere that explain how to setup the cluster, both with and without auth.   These will need updating.

I was considering how to document this feature, but at the moment, there is no way to use system (e.g. heartbeat, lag) REST calls with auth.  Unfortunately, this is the case, even without this PR.  The system calls like heartbeating have no credentials because they're background threads.  Without biting off the task of implementing system call authentication, they can't currently work without whitelisting (`authentication.skip.paths`) the particular system endpoints as being unauthenticated.  I'm not sure if you're fine with documenting/encouraging that setup, in this interim period before it's implemented.  Obviously, forwarded pull-queries, for example, still benefit from the authentication because they originate with a user.

I'm not sure if you meant just documenting auth in a cluster in general.

Is `ksql/docs/operate-and-deploy/installation/server-config/security.md` the document you meant?

> Going forward, we'll likely need a better story around auth on the internal listeners. This may be better done as a KLIP initially. Likely, users won't want the same auth mech on the internal as the external.

I agree.  In fact, we may want to do this outright.  Let's chat about this a bit to scope some of it out at a high level.

> 
> I think the current state is:
> 
> * Authentication:  currently ksql supports SSL, BASIC and/or a plugable mech.  Some paths can be configured to skip authentication.
> * Authorisation: ksql doesn't currently internally do authorisation, and uses its own credentials when talking to external services.  Though in the future we want ksql to be able to impersonate users when talking to external services.
> 
> With this change the above will also apply to the internal interface too. However, I'm not sure how useful that is! This shared auth mech would mean anyone that can call the public interface can also call the internal one. A more common pattern would be to allow the auth mech to be different on the two.

I don't think we'd ever want a system endpoint to be exposed, by definition.  If we wanted an endpoint that was open to some users, but not others, that would just be a question of authorization.

> The job of the inter-node authentication should be limit communication on the internal interface to trusted ksqlDB nodes. It should not be to authenticate users - that should be done only on the boundary to the ksqlDB cluster.

As you're implying, if the nodes were authenticated to each other, on an internal call, they could trust that another node had already authenticated the end user and accept their claim.

Thanks for all of the great suggestions.  I did everything but some of the documentation, because I have some followup questions that would be great to discuss over zoom.

purplefox commented:
Hey @AlanConfluent - sorry for the delay in getting back to this.

It's my understanding the reason for this PR is to create a new class of listeners which are to be used for internal endpoints. The PR wants this for two reasons:
1. To drop auth on those endpoints
2. To be able to configure those endpoints on a local address to prevent internal requests being routed through the public IP which might be a longer path.

As per offline discussions, given that we agree that internal endpoints should be authed (as discussed offline with @ijuma and @apurvam ) then, aiui, we only need a mechanism to allow internal calls to be routed directly to listeners on the local network. I think the current advertised.listeners functionality allows that to be done already, which would imply there is nothing we need to do here?

AlanConfluent commented:
> Hey @AlanConfluent - sorry for the delay in getting back to this.
> 
> It's my understanding the reason for this PR is to create a new class of listeners which are to be used for internal endpoints. The PR wants this for two reasons:
> 
> 1. To drop auth on those endpoints

In the current codebase, without my PR, the endpoints /heartbeat and /lag  don't work with auth.  You can whitelist them to avoid auth using `authentication.skip.paths`, but then you've exposed these internal endpoints, since they use the same listener as the public API.

With this PR, you can at least put them on a different listener, and separate calls from outside and inside your cluster.  This can allow for port filtering as a first step of authentication.  As a followup, we will also add authentication intra cluster.

So, to address this reason, I would say that auth doesn't exist on internal calls at the moment.  It is broken.

> 2. To be able to configure those endpoints on a local address to prevent internal requests being routed through the public IP which might be a longer route.

Correct.

> 
> As per offline discussions, given that we agree that internal endpoints should be authed (as discussed offline with @ijuma and @apurvam ) then, aiui, considering that internal endpoints are currently authed, we only need a mechanism to allow internal calls to be routed directly to listeners on the local network.
> 
> I think the current advertised.listeners functionality allows that to be done already, which would imply there is nothing we need to do here?

As I mentioned, 2) isn't the only purpose.  The internal system endpoints don't work with auth and are effectively broken.

> 
> (As an improvement to the way internal endpoints are currently authed: Right now they are authed with the creds of the public user. If instead, we authed internal endpoints with the creds of a special "server" user, and if we cache authorisations for some time on the server, then authing internal calls effectively becomes a performance no-op. We can deploy a long lived token for the server user creds when ksqlDB is provisioned in cloud. That way we still have auth, so don't reduce our security, but we can also do this in a performant way. Wdyt?)

I agree with this sentiment that we should have different roles, system and user, and Andy convinced me that we need to do mutual ssl auth.  We can discuss a bit more which route to go for intra-cluster auth, but hopefully that's a followup from this PR.

AlanConfluent commented:
> documentation: we have docs somewhere that explain how to setup the cluster, both with and without auth. These will need updating.

As we chatted about, I don't think this PR touches much on auth, except that I've whitelisted `/lag` and `/heartbeat` to be non auth until we implement intra node auth for system calls.  By default, these endpoints are disabled.

I've documented how to set up a ksqlDB cluster, including the use of `ksql.internal.listener`. 
 @JimGalasyn please tell me if these docs seem reasonable.

@big-andy-coates I think I've addressed your comments.
@purplefox  I hope you understand my motivation for this.  Happy to chat further if you disagree.

purplefox commented:
> So, to address this reason, I would say that auth doesn't exist on internal system calls at the moment. It is broken.

Could you elaborate a bit what you mean here? Those endpoints are definitely protected by auth and if auth is enabled on ksqlDB they can't be accessed without being authed. I worked on the porting of auth to Vert.x recently myself, and part of the work was making sure this is the case. So I'm a bit confused when you say it is broken.

AlanConfluent commented:
> > So, to address this reason, I would say that auth doesn't exist on internal system calls at the moment. It is broken.
> 
> Could you elaborate a bit what you mean here? Those endpoints are definitely protected by auth and if auth is enabled on ksqlDB they can't be accessed without being authed. I worked on the porting of auth to Vert.x recently myself, and part of the work was making sure this is the case. So I'm a bit confused when you say it is broken.

I mean that the auth checks are there and the system endpoints are protected when auth is enabled, but then when the system tries to issue the calls itself, it will fail because it has no credentials (since they're background threads and have no user they're working on behalf of).

So, the endpoints work with no auth, or fail when auth is enabled.  The handlers exist, but they don't work as intended.

purplefox commented:
> > > So, to address this reason, I would say that auth doesn't exist on internal system calls at the moment. It is broken.
> > 
> > 
> > Could you elaborate a bit what you mean here? Those endpoints are definitely protected by auth and if auth is enabled on ksqlDB they can't be accessed without being authed. I worked on the porting of auth to Vert.x recently myself, and part of the work was making sure this is the case. So I'm a bit confused when you say it is broken.
> 
> I mean that the auth checks are there and the system endpoints are protected when auth is enabled, but then when the system tries to issue the calls itself, intra-node (e.g. heartbeating), it will fail because it has no credentials (since they're background threads and have no user they're working on behalf of).
> 
> So, the endpoints work with no auth, or fail when auth is enabled. The auth handlers exist for these endpoints, but they don't work as intended.

Thanks for the update Alan. I think the endpoints are working fine - they are requiring auth, but the request is not providing credentials so they are being rejected. This is expected bahaviour.

Imho, when any new feature is added (e.g. /lag and /heartbeat) we need to make sure that feature works with our auth model. Currently many of our users use auth either in ksqlDB standalone, or in CP, and cloud runs with auth too, so it's important the feature doesn't only work if the user is running without auth.

AlanConfluent commented:
I believe we have an agreement about how to go forward.  I've removed the system endpoints from the hardcoded auth whitelist `KsqlAuthorizationProviderHandler.PATHS_WITHOUT_AUTHORIZATION`, so that we're conservative by default, requiring auth for all internal endpoints.  In order to use port filtering with this, you must use the config `authentication.skip.paths` to allow their use within a private network.  I've explained this in docs.

As part of the followup work, I will do intra-cluster authentication, likely with SSL mutual auth.  Will send out a PR for that soon after this.

Now that we're in agreement at a high level, please review the code! :-)

purplefox commented:
>> As part of the followup work, I will do intra-cluster authentication, likely with SSL mutual auth.

As a note of caution - creating a connection each time for inter node traffic is already slow, requiring TLS on it will make it even slower (TLS handshake requires 2 network round trips), so if you go down this route I'd recommend not doing this until we are caching inter node connections.

Another way of doing this that doesn't require auth that you might want to consider is to whitelist source IPs when connecting to internal endpoints.

E.g. you only accept connections to internal endpoints if source IP is in the list of known server IPs (these can be configured by the cloud provisioning process). Alternatively wildcards could be used, e.g. saying only accept connections from source IP 192.168.1.*. This might be a bit simpler than implementing auth (whether it's via token or certificate based).

big-andy-coates commented on docs/operate-and-deploy/installation/server-config/index.md:
```diff
@@ -310,3 +310,60 @@ Start the ksqlDB server with the configuration file specified.
+
+### Internal Network
+
+For the typical setup within an internal network, you might have just one
```
I'm a bit confused by the 'internal network' terminology, though it could just be me.

'Internal network' suggests to me that there is an internal and public network, which is more an IaaS setup.  Yet the example is talking more about an example where the `listeners` config is either using some form of `localhost` for the `listeners` config, or where the local host name is not resolvable/routable by other servers running the ksql nodes.

Is this just me? Or would a better name help here?  It seems we have the following setups:

1. Single interface, using resolvable and routable host names, i.e. just using `listeners` config.
2. Single interface, using unresolvable or unroutable host names, i.e. using `listeners` and `advertised` configs.
3. Dual interface, where internal is resolvable and routable host name, i.e. using `listeners` and `internal.listeners`.
4. Dual interface, where internal is unresolvable or unroutable host name, i.e. using `listeners`, `internal.listeners` and `advertised.listeners`.

I think it would make sense to have a section for each and explain the key part of needing to set advertised config is if either:
  - the config used for internal listeners, (either set explicitly or defaulting to `listeners[0]`), is using wildcards (which defaults to localhost) or explicitly using localhost, i.e. so it's not meaningful to other nodes
   - or the explicit host name used for internal listeners, (either set explicitly or defaulting to `listeners[0]`), is either not resolvable to an IP, or that IP is not routable, from other ksql nodes.

The choice between these 4 modes doesn't have that much to do with the users choice between running auth or not running auth, (as you've alluded to in the Unsecure Environment section below). After all, they can run each with auth on or off.   However, _which_ of these configurations they're running may affect how they need to configure auth. Maybe not right now, but certainly once/if we allow different auth to be set on the internal endpoints.  Hence I'd probably change the text in this page to say that running over dual network interfaces makes it possible to configure the network such that the inter-node endpoints ksqlDB uses are not accessible on the public listener(s).  If done correctly, this can be a more secure mode of deployment. And then go one to briefly mention and link to the auth docs as a 'Another security feature you can use is Auth'.

AlanConfluent replied:
I agree that the previous terminology wasn't great.

I swapped out what I had for 4 configurations, as you've outlined above.  I also mentioned the fallback sequence, so it should be intuitive what's in use when configurations are unset.  I kept things fairly simple, so I didn't draw the full decision tree, but I think for documentation, we want to keep it somewhat straightforward.

I also mentioned the dual network interfaces and how this is a similar variation to port filtering.  I agree that this is going to need to be elaborated on a fair amount when we add inter-node auth and this blurb will probably have to be updated as well to at least reference it.

big-andy-coates commented on ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/AuthorizationFunctionalTest.java:
```diff
@@ -121,6 +125,33 @@ public void shouldAllowAccess() {
   }
 
+  @Test
+  public void shouldDenyAccess_internal() {
```
perfect - thanks!

AlanConfluent commented:
> TLDR; Verticles are designed as a way of scaling a server, and you typically don't want lots of them. They're not designed as units of code encapsulation. It's fine to encapsulate your routing code across (potentially) many classes, but usually you just want to deploy a single router and N verticles where N roughly matches number of cores on the machine.

I use just the one verticle now, as you suggested.

> 
> If you create a handler that rejects requests to internal endpoints if they're not received on an internal listener, e.g. something like https://gist.github.com/purplefox/99d387a4c8ebafdc8fe7ca37a819cd7f - you can place that handler before the all the internal endpoints, which would just be a few lines of extra code.

Thanks for this very concrete pointer.  I pretty much adopted that.  I modified it by taking note of the listener I install it on (whether it's internal or not), and just pass that it, so I don't have to look at the local address and match it up against the listener.  Comparing host names could get slightly error prone (given 0.0.0.0 and that fact that ServerAddress only exposes host() which sometimes can be an address and other times a hostname).

> This also means you probably don't need ksql.internal.listeners any more as this will work with ksql.advertised.listeners which should help reduce config overload.

I think I still need this config.  You might want to bind https://0.0.0.0:9188 but advertise some other logical hostname that resolves to you like https://host-1.internal.example.com:9188.  It's easier than binding the exact ip address which is often abstracted away/changing to some degree in various environments.

> The handler approach could be extended in the future to also reject _source_ IPs if they don't correspond to whitelisted IPs of other servers in the cluster, or don't match a wildcard (e.g. 192.168.1.*). That might be a way of getting better security for internal endpoint access without having to require auth. Just a thought!

Good idea.  I'm going to ask more about this.

AlanConfluent commented:
> Am I right in thinking that with these internal paths removed from `PATHS_WITHOUT_AUTHORIZATION` then these features won't work if the user turns on auth? Or will it depend _what_ auth is turned on? e.g. if using mutual SSL then it would work, but if using BASIC / Jaas then it wouldn't? We need to be careful we don't end up with a really poor UX here if a user is using these features and turns on the common forms of auth.

If they don't also set `authentication.skip.paths`, then yes, the features won't work if BASIC / Jaas is enabled.  I agree that it doesn't really make sense since they are designed not to work with those forms of auth at the moment.  It will require a future change, enabling intra-cluster authentication to have them play nicely, which is coming soon.

@purplefox felt strongly that we shouldn't add them to `PATHS_WITHOUT_AUTHORIZATION` since this would expose these endpoints (if they were enabled) without auth until we can add intra-cluster authentication.

I agree that the UX isn't great at the moment but it was in the name of security.

> > I think technically, `ksql.advertised.listeners` will still default to listeners[0]. Maybe it should do as you're saying and fall back to `internal.listener` before that, but I wanted to keep it easy to reason about.
> 
> `internal.listener` is the listener used for inter-node communication. `ksql.advertised.listeners` is the url shared between nodes for inter-node communication: it's not used for anything else. So the way it is at the moment is surely wrong. If I configure ksql with a different `internal.listener` to `listeners[0]`, and don't set `advertised.listeners`, then it will incorrectly advertise the public listener for inter-node comms, where as it should be advertising the internal listener!

Fair point, I'll change this to fall back on `internal.listener`.

AlanConfluent commented on docs/operate-and-deploy/installation/server-config/config-reference.md:
```diff
@@ -415,6 +415,30 @@ The corresponding environment variable in the
+This address is used to initiate intra-cluster requests such as forwarded pull queries,
+heartbeating, and lag reporting.
+
+If not set, the system will use the first hostname in `listeners` to advertise, and if this
+is set to a wildcard address, the system hostname is used.
```
I made it behave this way and removed this explanation anyway, per your previous comment.

AlanConfluent commented on docs/operate-and-deploy/installation/server-config/config-reference.md:
```diff
@@ -415,6 +415,30 @@ The corresponding environment variable in the
 
+### ksql.advertised.listener
+
+The `ksql.advertised.listener` setting controls the address that is advertised within
+{{ site.kstreams }} and shared with other ksqlDB nodes that make up the cluster.
+This address is used to initiate intra-cluster requests such as forwarded pull queries,
+heartbeating, and lag reporting.
```
Done

AlanConfluent commented on docs/operate-and-deploy/installation/server-config/index.md:
```diff
@@ -310,3 +310,60 @@ Start the ksqlDB server with the configuration file specified.
+Configuring Listeners for a ksqlDB Cluster
+--------------------------------
+
+Multiple hosts are required to scale ksqlDB processing power and to do that, they must
+form a cluster.  ksqlDB utilizes {{ site.kstreams }} to do this and requires all hosts of
+a cluster to use the same `ksql.service.id`.
```
Removed.

AlanConfluent commented on docs/operate-and-deploy/installation/server-config/index.md:
````diff
@@ -310,3 +310,60 @@ Start the ksqlDB server with the configuration file specified.
+ksql.advertised.listener=http://host1.internal.example.com:8088
+```
+
+### Unsecured Environment
````
I think I meant originally that if you have an `Untrusted Environment`, you should take security measures.  I agree this is rather confusing.

Going with your suggestion

AlanConfluent commented on ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/Server.java:
```diff
@@ -224,6 +277,10 @@ public int queryConnectionCount() {
 
+  public synchronized List<URI> getInternalListeners() {
+    return ImmutableList.copyOf(internalListeners);
+  }
```
You're right.  I was originally considering allowing for more than one, though I didn't do it this way.  I changed the return type to an Optional<URI> which is immutable.

AlanConfluent commented on ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestConfig.java:
```diff
@@ -137,6 +137,16 @@
+          + "This is the same endpoint referred to by ksql.advertised.listener, which is also used "
+          + "for internal requests."
+          + "In IaaS environments, this may need to be different from the interface to which"
+          + "the other endpoints bind.";
```
I agree, those previous doc strings were a bit outdated after multiple iterations on this.

Adopted your suggestions.

I added more of a security explanation for the internal listener doc, which I think makes it easier to understand the purpose of.

AlanConfluent commented:
I reverted the refactoring of `ServerVerticle` at Tim's request.  This will be split into another change.

Ready for review again @purplefox

purplefox commented on ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/InternalEndpointHandler.java:
```diff
@@ -0,0 +1,45 @@
+  public void handle(final RoutingContext routingContext) {
+    if (INTERNAL_PATHS.contains(routingContext.normalisedPath())
+        && !isFromInternalListener) {
+      routingContext.fail(400,
```
Best to use constant here HttpResponseStatus.BAD_REQUEST.code()

AlanConfluent replied:
Done
